client/src/contexts/user/UserContextProvider.js

```diff
@@ -0,0 +1,107 @@
+import * as React from "react";
+import { SET_USER, LOGOUT } from "../types";
+
+
+
+const initialState = {
+    isAuthenticated: false,
+    profile: null,
+};
+
+const UserReducer = (state, action) => {
+    switch (action.type) {
+        case SET_USER:
+            return {
+                ...state,
+                isAuthenticated: true,
+                profile: action.payload.user,
+            };
+        case LOGOUT:
+            return {
+                ...state,
+                isAuthenticated: false,
+                profile: null,
+            };
+        default:
+            return state;
+    }
+};
+
+const UserContext = React.createContext(initialState);
+
+const UserContextProvider = ({ children }) => {
+    const [state, dispatch] = React.useReducer(UserReducer, initialState);
+
+    const register = async (formValues) => {
+        const response = await fetch("/auth/register", {
+            method: "post",
+            body: JSON.stringify(formValues),
+            headers: {
+                "Content-Type": "application/json",
+            },
+        });
+
+        const data = await response.json();
+
+        if (!data.user) {
+            return {
+                result: false,
+                message: data.errors,
+            };
+        } else {
+            dispatch({ type: SET_USER, payload: data });
+            return { result: true };
+        }
+    };
+
+    const login = async (formValues) => {
+        const response = await fetch("/auth/login", {
+            method: "post",
+            body: JSON.stringify(formValues),
+            headers: {
+                "Content-Type": "application/json",
+            },
+        });
+
+        const data = await response.json();
+
+        if (!data.user) {
+            return {
+                result: false,
+                message: data.errors,
+            };
+        } else {
+            dispatch({ type: SET_USER, payload: data });
+            return { result: true };
+        }
+    };
+
+    const checkLogin = async () => {
+        const response = await fetch("/auth/user", {
+            method: "get",
+            credentials: 'include',
+        });
+
+        const data = await response.json();
+
+        if (data.user) {
+            dispatch({ type: SET_USER, payload: data });
+        }
+    };
+
+    React.useEffect(() => {
+        if(document.cookie.indexOf('token') === -1){
+            checkLogin().catch((error) => {
+                console.log(error)
+            })
+        }
```

Comment:
ok, so a couple thoughts:
- we want to call checkLogin() regardless of the cookie status. Because we're using httpOnly, we won't be able access the cookie using document.cookie (https://javascript.info/cookie#httponly)

- checkLogin is async, so if we await it that might fix the flicker. Async inside a useEffect can be tricky; you'll need to do it like:
```
useEffect(() => {
    const checkCookie = async () => await checkLogin();
    checkCookie();
}, [ ] )
```

Reply:
Thanks! I've made the updates but there is still a flicker.  Does it have to do with redirects in the routes?

Reply:
Yeah, it's probably something to do with the redirects. Try moving the useEffect to App.js

Reply:
I've tried that and it still flickers. Can I go head and complete the pull request?